Ху Юйхан.sql

```diff
@@ -0,0 +1,102 @@
+/***
+Проблема 1 Анализ:
+Из описания мы знаем, что dbo.F_WORKS_LIST() - это определяемая пользователем функция (UDF), 
+которая может быть предназначена для возврата списка заказов. Хотя у нас нет реального сценария для 
+этой функции, мы можем определить возможные недостатки и проблемы производительности, 
+основываясь на обычном использовании UDF в SQL Server и соображениях производительности.
+Выявленные недостатки и потенциальные проблемы с производительностью:
+UDF как источник данных: использование UDF непосредственно в операторах SELECT (особенно если 
+они возвращают большое количество строк) может привести к проблемам с производительностью. 
+Каждый вызов UDF выполняет свою собственную внутреннюю логику SQL, и эта логика может быть 
+повторно выполнена для каждой строки возвращаемого результата.
+Отсутствие фильтрации и сортировки: использование SELECT TOP 3000 * может означать отсутствие 
+фильтрации по определенным условиям и сортировки результатов. Это может привести к возврату 
+ненужных данных и увеличению времени передачи данных по сети и обработки на стороне клиента.
+Внутренняя логика UDF: если UDF содержит сложную внутреннюю логику запросов, такую как 
+соединения нескольких таблиц, подзапросы, агрегатные функции и т. д., это может существенно 
+повлиять на производительность.
+Отсутствие индексов: если таблица Works (или связанные с ней таблицы) не имеет надлежащих индексов 
+для поддержки запросов в UDF, это может повлиять на производительность.
+Борьба за ресурсы: если UDF вызывается часто в часы пик, он может конкурировать с другими запросами 
+за ресурсы процессора, памяти или ввода-вывода, что приведет к снижению производительности.
+Размер набора результатов: SELECT * может возвращать большой объем данных, особенно если таблица 
+имеет много столбцов. Это может увеличить время передачи данных по сети и время обработки запроса 
+клиентом.
+***/
+
+
+--Вторая задача：
+ALTER FUNCTION [dbo].[F_WORKS_LIST]()
+RETURNS @Results TABLE (
+ ID_WORK INT,
+ ResponsibleEmployeeFullName VARCHAR(101),
+ NotCompletedWorkItemsCount INT,
+ CompletedWorkItemsCount INT
+)
+AS
+BEGIN
+ -- 1. 获取所有 Works 和对应的 EmployeeID （ Получите все работы и соответствующий
+идентификатор сотрудника）
+ DECLARE @WorksCursor CURSOR FOR
+ SELECT TOP 3000 ID_WORK, EmployeeID -- 限制检索的订单数量为 3000（Ограничьте количество
+полученных заказов до 3000）
+ FROM Works;
+ -- 打开游标并处理每个 Works 条目（Откройте курсор и обработайте каждую запись о работе）
+ OPEN @WorksCursor;
+ FETCH NEXT FROM @WorksCursor INTO @ID_WORK, @EmployeeID;

```

Comment:
Msg 137, Level 15, State 2, Procedure F_WORKS_LIST_OPT2, Line 16 [Batch Start Line 0]
Must declare the scalar variable "@ID_WORK".